src/reflect/scala/reflect/internal/transform/Erasure.scala

```diff
@@ -112,8 +112,9 @@ trait Erasure {
     protected def eraseDerivedValueClassRef(tref: TypeRef): Type = erasedValueClassArg(tref)
 
     def apply(tp: Type): Type = tp match {
-      case ConstantType(_) =>
-        tp
+      case ConstantType(ct) =>
+        if (ct.tag == ClazzTag) ConstantType(Constant(apply(ct.typeValue)))
+        else tp
```

Comment:
I think this broke Scala.js. In *some* cases, a `classOf[Unit]` becomes `classOf[BoxedUnit]` after erasure in 2.12.0-M4, but it did not up and until March 31, before this PR was merged (it stayed a `classOf[Unit]`). The "some" cases include when compiling `java.lang.Void` and `scala.runtime.BoxedUnit` in Scala.js; but does not include when compiling the hello world or the test suite. I did not try to minimize this, so far :-s

This causes the following 3 Scala.js tests to fail:

* org.scalajs.testsuite.javalib.lang.ClassTest
* org.scalajs.testsuite.scalalib.ArrayBuilderTest
* org.scalajs.testsuite.scalalib.ClassTagTest

See the full log here: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/364/consoleFull (grep for `[scala-js] [error] Failed tests:`).

At least for `ClassTest`, this happens because `java.lang.Void.TYPE` and `scala.runtime.BoxedUnit.TYPE` return the `class scala.runtime.BoxedUnit` instead of `void`. And that, because, when those classes were compiled, `classOf[Unit]` was transformed into `classOf[BoxedUnit]` during erasure. The source for these files are at https://github.com/scala-js/scala-js/blob/master/javalanglib/src/main/scala/java/lang/Void.scala and https://github.com/scala-js/scala-js/blob/master/library-aux/src/main/scala/scala/runtime/BoxedUnit.scala

Any idea?